patroni/postgresql/validator.py

```diff
@@ -128,456 +135,370 @@ def transform(self, name: str, value: Optional[Any]) -> Optional[Any]:
+            section_var = getattr(module, section)
+
+            config_section = config.get(section, {})
+            for parameter in config_section.keys():
+                for exc in _load_postgres_guc_validators(section_var, config_section, parameter):
+                    logger.warning('Faced an issue while parsing a validator for parameter `%s`, from file `%s`: `%s`',
+                                   parameter, file, str(exc))
+
+
+_load_postgres_gucs_validators()
 
 
 def _transform_parameter_value(validators: MutableMapping[str, Union[_Transformable, Tuple[_Transformable, ...]]],
-                               version: int, name: str, value: Any) -> Optional[Any]:
-    name_validators = validators.get(name)
-    if name_validators:
-        for validator in (name_validators if isinstance(name_validators, tuple) else (name_validators,)):
-            if version >= validator.version_from and\
-                    (validator.version_till is None or version < validator.version_till):
-                return validator.transform(name, value)
+                               version: int, name: str, value: Any,
+                               available_gucs: CaseInsensitiveDict) -> Optional[Any]:
+    """Validate *value* of GUC *name* for Postgres *version* using defined *validators* and *available_gucs*.
+
+    :param validators: a dictionary of all GUCs across all Postgres versions. Each key is the name of a Postgres GUC,
+        and the corresponding value is a tuple of tuples. Each sub-tuple is a validation rule for the GUC for a given
+        range of Postgres versions. Should either contain recovery GUCs or general GUCs, not both.
+    :param version: Postgres version to validate the GUC against.
+    :param name: name of the Postgres GUC.
+    :param value: value of the Postgres GUC.
+    :param available_gucs: a dictionary of all GUCs available in Postgres *version*. Each key is the name of a Postgres
+        GUC, and the corresponding value is irrelevant. Used for a couple purposes:
+        * Disallow writing GUCs to ``postgresql.conf`` (or ``recovery.conf``) that does not exist in Postgres *version*;
+        * Avoid ignoring GUC *name* if it does not have a validator in *validators*, but is a valid GUC in Postgres
+            *version*.
+
+    :returns: the return value may be one among:
+        * *value* transformed to the expected format for GUC *name* in Postgres *version*, if *name* is present in
+            *available_gucs* and has a validator in *validators* for the corresponding Postgres *version*; or
+        * The own *value* if *name* is present in *available_gucs* but not in *validators*; or
+        * ``None`` if *name* is not present in *available_gucs*.
+    """
+    if name in available_gucs:
+        name_validators = validators.get(name)
+        if name_validators:
+            for validator in (name_validators if isinstance(name_validators, tuple) else (name_validators,)):
```

Comment:
Isn't it now always a tuple (due to https://github.com/zalando/patroni/pull/2671/files#diff-c1f5f15afbae52f3272b8520d30aea67a3820c6855b92a90ed5494f27fdd1ec3R258)?